components/builder-jobsrv/src/server/worker_manager.rs

```diff
@@ -865,15 +848,38 @@ impl WorkerMgr {
 
     fn process_job_status(&mut self) -> Result<()> {
         self.rq_sock.recv(&mut self.msg, 0)?;
+
+        let worker_name = self.msg.as_str();
+        debug!("Received job status from worker {:?}", &worker_name);
+
         // TODO get worker id here from msg
         let worker_id = "Unknown".to_owned();
 
         self.rq_sock.recv(&mut self.msg, 0)?;
 
+        //  We will get a job or a heartbeat.
         let job = Job::new(Message::parse_from_bytes(&self.msg)?);
         debug!("Got job status: {:?}", job);
         self.datastore.update_job(&job)?;
 
+        //  Create and process the heartbeat from job data.
+        let mut heartbeat = jobsrv::Heartbeat::new();
+        heartbeat.set_endpoint(job.get_worker().to_string());
+        let worker_state = match job.get_state() {
+            jobsrv::JobState::Dispatched
+            | jobsrv::JobState::Pending
+            | jobsrv::JobState::Processing
+            | jobsrv::JobState::CancelPending
+            | jobsrv::JobState::CancelProcessing => jobsrv::WorkerState::Busy,
+            jobsrv::JobState::Rejected
+            | jobsrv::JobState::CancelComplete
+            | jobsrv::JobState::Complete
+            | jobsrv::JobState::Failed => jobsrv::WorkerState::Ready,
+        };
+        heartbeat.set_state(worker_state);
+        heartbeat.set_target(job.get_target().to_string());
+        self.process_heartbeat(&heartbeat)?;
```

Comment:
maybe have `process_job_status` return a `Result<Heartbeat>` and then have the caller call `process_heartbeat`. This feels a little tangled.